tests/test_all_sensors.py

```diff
@@ -0,0 +1,313 @@
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import logging
+import math
+import unittest
+from dataclasses import dataclass
+
+import numpy as np
+import pytest
+from lsst.ts.ess import common
+
+logging.basicConfig(
+    format="%(asctime)s:%(levelname)s:%(name)s:%(message)s", level=logging.DEBUG
+)
+
+# Data for the Campbell CSAT3B sensor from page 55 of csat3b.pdf.
+CSAT3B_DATA = [
+    "0.08945,0.06552,0.05726,19.69336,0,5,c3a6",
+    "0.10103,0.06517,0.05312,19.70499,0,6,3927",
+    "0.09045,0.04732,0.04198,19.71161,0,7,d7e5",
+    "0.08199,0.03341,0.03421,19.73416,0,8,4ad9",
+    "0.08867,0.03522,0.03378,19.75360,0,9,e314",
+    "0.08675,0.02142,0.03289,19.76858,0,10,9b60",
+    "0.09035,0.01987,0.03667,19.78433,0,11,931a",
+    "0.09960,0.02615,0.04330,19.79236,0,12,14a1",
+    "0.09489,0.02513,0.05120,19.79083,0,13,0c0d",
+    "0.09513,0.02403,0.05648,19.79037,0,14,c30d",
+    "0.10715,0.02723,0.05739,19.78729,0,15,a14c",
+    "0.11630,0.03674,0.05579,19.78812,0,16,5cD7",
+]
+
+# Data for the wind sensor.
+WIND_DATA_LIST = [
+    ["015.00", "010"],
+    ["001.00", ""],
+    ["015.00", "010"],
+]
+
+# Alias for the type of the sensor data.
+SensorDataType = dict[str, common.TelemetryDataType]
+
+
+@dataclass
+class SensorAndData:
+    sensor: common.sensor.BaseSensor
+    data: SensorDataType
+
+
+class AllSensorsTestCase(unittest.IsolatedAsyncioTestCase):
+    async def verify_sensor_telemetry(self) -> None:
+        """Verify the sensor telemetry.
+
+        Loop over a list of sensors and loop over the provided data to verify
+        that the sensor performs as expected.
+        """
+        for sensor_and_data in self.sensor_and_data_list:
+            sensor = sensor_and_data.sensor
+            data = sensor_and_data.data
+            for raw_telemetry in data:
+                reply = await sensor.extract_telemetry(line=raw_telemetry)
+                for index, expected_reply in enumerate(data[raw_telemetry]):
+                    if isinstance(expected_reply, float):
+                        if math.isnan(expected_reply):
+                            assert math.isnan(reply[index])
+                        else:
+                            assert math.isclose(
+                                expected_reply, reply[index], abs_tol=0.005
+                            )
+                    else:
+                        assert expected_reply == reply[index]
```

Comment:
Please consider to refactor to decrease the indentation level.

Reply:
Done